### PR TITLE
fix(jetbrains): avoid connection loops

### LINKED ISSUE #9848: Refresh workspace page or revisit it and you end up in jetbrains spawning hell
### Bug description

If you leave the workspace page and return to it, and, I beleive also just refresh the page. It starts spawning new instances to which there is no escaping.

### Steps to reproduce

Leave open workspace page, return to open workspace page 

Screenshot 2022-05-08 015708

### Workspace affected

_No response_

### Expected behavior

_No response_

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andreafalzetti**
## Description
Adds a check to make sure to avoid connecting multiple times to the same workspace. Not having this check in place is causing a loop of connections when trying to open a connection with a workspace that is already opened in another client.

The approach used is to show this message popup while we wait for a new API to request focus to the open IDE.

Screenshot 2022-09-01 at 23 37 26

## Related Issue(s)

Fixes #9848

## How to test
1. Ensure you have the [latest JetBrains Gateway](https://www.jetbrains.com/remote-development/gateway/) installed.
2. Download the plugin build related to this branch in [Dev Versions](https://plugins.jetbrains.com/plugin/18438-gitpod-gateway/versions/dev), and [install it on the Gateway](https://www.jetbrains.com/help/idea/managing-plugins.html#install_plugin_from_disk).
3. Start a workspace in gitpod.io with any JetBrains IDE as your preferred editor
4. Open the IDE as you normally do
5. Try launching new sessions by clicking on the "Open in IntelliJ" (or GoLand, etc)

Optionally,

1. Open DevTools on the workspace browser page where you have the open in intellij button
2. Simulate quick clicks on the button:
```
document.querySelectorAll('button')[1].click();
document.querySelectorAll('button')[1].click();
document.querySelectorAll('button')[1].click();
```
3. Observe how 3 popups open but the main client does not stop or if affected.

## Release Notes

```release-note
Fixed JetBrains connection loop when connecting twice to the same workspace 
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@andreafalzetti: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt [hunk 1]
```diff
@@ -74,6 +76,18 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
             parameters["workspaceId"]!!,
             parameters["backendPort"]
         )
+
+        val existingConnection = activeConnections.find {
+            it.gitpodHost == parameters["gitpodHost"] && it.workspaceId == parameters["workspaceId"] && it.backendPort == parameters["backendPort"]
+        }
+
+        if (existingConnection != null) {
```
Comment (akosyakov) [inline components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:84]: Java is concurrent you will need to synchronize this place, you can use concurrentmap it has methods which already synchronize check out other places for instance how we ensure that only connection to server api exists

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt [hunk 2]
```diff
@@ -230,6 +244,12 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                                         setErrorMessage("failed to fetch JetBrains Gateway Join Link.")
                                         return@launch
                                     }
+
+                                    // TODO: is this the right place?
```
Comment (akosyakov) [inline components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:248]: It should happen in the check above, ie in one transaction

Comment (andreafalzetti) [reply components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:254]: @akosyakov could you clarify what you mean for `transaction` in this context?

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt [hunk 3]
```diff
@@ -144,6 +148,22 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
...
+            val title = "Oops"
+            val okButton = Messages.getOkButton()
+            val options = arrayOf(okButton)
+            val defaultIndex = 0
+            val icon = Messages.getWarningIcon()
+            Messages.showDialog(message, title, options, defaultIndex, icon)
+
+            val errMessage = "A connection to the same workspace already exists: $connectionKeyId"
+            thisLogger().warn(errMessage)
+            throw IllegalArgumentException(errMessage)
```
Comment (akosyakov) [inline components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:164]: IllegalArgumentException - means that invalid argument was passed

maybe IllegalStateException is better here, not sure

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt (hunk 2 above)
Comment (andreafalzetti) [reply components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt]: Thanks @felladrin for clarifying in our sync chat, I think I understand what @akosyakov meant and I've updated the code accordingly. 🙏

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt [hunk 4]
```diff
@@ -144,6 +148,24 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
             connectionPanel.repaint()
         }
 
+        val connectionKeyId = "${connectParams.gitpodHost}-${connectParams.workspaceId}-${connectParams.backendPort}"
+
+        if (activeConnections.containsKey(connectionKeyId)) {
```
Comment (andreafalzetti) [inline components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:153]: This should be thread-safe to my understanding, using `putIfAbsent`.

I've tried other approaches like `getOrPut` and `computeIfPresent` but with that approach, I didn't find a way to have an "else" condition.

File: components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt [hunk 5]
```diff
@@ -216,7 +245,8 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                             thinClientJob = launch {
                                 try {
                                     val updatedIdeUrl = URL(update.ideUrl);
-                                    val sshHostUrl = URL(update.ideUrl.replace(update.workspaceId, "${update.workspaceId}.ssh"));
+                                    val sshHostUrl =
```
Comment (andreafalzetti) [inline components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt:248]: formatted by IntelliJ

**Review by akosyakov — APPROVED** [review]
code-wise it looks great

/hold

@felladrin Could you help with testing please? 🙏

**Comment by felladrin** [comment]
Sure! I'll test it today soon.

**Comment by felladrin** [comment]
ℹ️ I'm reviewing it now!

**Comment by felladrin** [comment]
Restarting the Preview Environment, as it expired.

/werft run

**Review by felladrin — APPROVED** [review]
Code reviewed and tested! ✅ 

Detection was really fast! Worked for all cases I tested.

image

**Comment by andreafalzetti** [comment]
thanks folks 🙏 

/unhold

**Pull request merged**